polar-core/src/partial/simplify.rs

```diff
@@ -1,217 +1,128 @@
+                (
+                    Value::Expression(Operation {
+                        operator: Operator::Dot,
+                        args,
+                    }),
+                    Value::Expression(Operation {
+                        operator: Operator::Dot,
+                        ..
+                    }),
+                ) => term.clone_with_value(Value::Expression(Operation {
+                    operator: Operator::Dot,
+                    args: vec![replacement.clone(), args.get(1).unwrap().clone()],
+                })),
+                _ => term.clone(),
             }
-        } else {
-            term.clone()
         }
-    } else {
-        term.clone()
-    }
-}
 
-fn get_roots(bindings: &Bindings) -> HashSet<Symbol> {
-    let mut roots = HashSet::new();
-    for (symbol, val) in bindings.iter() {
-        if !symbol.is_temporary_var() {
-            if let Value::Partial(_) = val.value() {
-                roots.insert(symbol.clone());
+        // Optionally sub `replacement` into each of the arguments of the operations.
+        let mut map_ops = |ops: &[Operation], replacement: &Term| -> TermList {
+            ops.iter()
+                .map(|o| Operation {
+                    operator: o.operator,
+                    args: o.args.iter().map(|a| sub_this(a, replacement)).collect(),
+                })
+                .map(|o| replacement.clone_with_value(Value::Expression(fold_operation(o, self))))
+                .collect()
+        };
+
+        match o.operator {
+            Operator::Unify => {
+                let left = o.args.get(0).unwrap();
+                let right = o.args.get(1).unwrap();
+                Operation {
+                    operator: Operator::And,
+                    args: match (left.value(), right.value()) {
+                        // Distribute expression over the partial.
+                        (Value::Partial(c), Value::Expression(_)) => map_ops(&c.operations, right),
+                        (Value::Expression(_), Value::Partial(c)) => map_ops(&c.operations, left),
+                        _ => return fold_operation(o, self),
+                    },
+                }
             }
+            _ => fold_operation(o, self),
         }
     }
-
-    roots
-}
-
-fn to_expressions(bindings: &mut Bindings) {
-    let mut new_bindings = Bindings::new();
-
-    for (name, val) in bindings.iter() {
-        if let Value::Partial(partial) = val.value() {
-            let name = name.clone();
-            let partial = partial.clone().into_expression();
-            new_bindings.insert(name, partial);
-        }
-    }
-
-    bindings.extend(new_bindings.into_iter());
 }
 
-fn remove_temporaries(bindings: &mut Bindings) {
-    let mut remove = HashSet::new();
-
-    for (name, _) in bindings.iter() {
-        if name.is_temporary_var() {
-            remove.insert(name.clone());
+/// Simplify a partial until quiescence.
+fn simplify_partial(mut term: Term) -> Term {
```

Comment:
Do we have any input cases where this actually uses the loop?

Reply:
`test_partial_nested_dot_ops` fails without the loop